### ISSUE TITLE
Clickhouse migrations and extensibility points

### ISSUE THREAD

**Issue by Ian2012** [openedx/tutor-contrib-clickhouse#22]
This issue aims to highlight the importance of adding support for ClickHouse migrations to enhance the development experience, facilitate efficient database schema evolution and maintenance, and provide bases for clickhouse customization in the context of the OARS project.

My current proposal is to create an openedx plugin that handles the migrations of clickhouse and facilitates the extraction of data from the clickhouse DB or use the existing [event-sink](https://github.com/openedx/openedx-event-sink-clickhouse) plugin, which will prevent us to use a clickhouse connection directly and use SQL statements.

It's important to provide a way to extend the clickhouse tables by plugins that aim to introduce extensibility to the data. It could just make sure that every table has a unique ID that allows developers to create relationships between custom tables or anything that allow developers to cover special use cases. 

The library used would be: [django-clickhouse-backend](https://pypi.org/project/django-clickhouse-backend/) as it's well maintained.

**Comment by bmtcril** [comment]
I agree that Django isn't well suited to this purpose, especially since the ClickHouse plugin doesn't seem to support some of the features we would definitely need like materialized views. Alembic seems somewhat better in this regard, so I'd consider going that way first. I definitely don't think schema management should be tied to edx-platform, we could just run migrations (either Alembic or the Cairn style) from the Tutor plugin. 

I believe that for your use cases you could create tracking log events in the frontend and/or backend to capture the data you want using either the existing tracking log to ClickHouse Vector pipeline or by creating an edx-platform plugin that extends event-routing-backends to transform those to xAPI and use the LRS or Vector pipelines to get them to ClickHouse. From there a custom dbt package can be created to make performant downstream tables and views for querying back through Superset or whatever other custom mechanism you need. I will say that what you get by using the Superset API is baked in security.

In general I agree that we shouldn't have a dependency on ClickHouse from the platform, but I think the clickhouse event sink shows a path where we don't include it as a python dependency and can still use it for the types of large scale real time enrichment data that LRSs and tracking logs really aren't designed for. I'm definitely open to other options if there's something better we can do that doesn't blow up the complexity of the project.
